### PR TITLE
FX vanilla option

### PR DESCRIPTION
**Pull request by yukiiwashita**
* Implementation of primary volatility providers for FX options: BlackVolatilitySmileFxProvider, BlackVolatilitySurfaceFxProvider, BlackVolatilityFlatFxProvider.
* Implementation of FxOptionSensitivity
* Implementation of FxVanillaOption, and its pricer, BlackFxVanillaOptionProductPricer. 

This work closes #314 
This work closes #317

### CODE REVIEW

**Comment by jodastephen** [comment]
Domain model parts merged into master. Branch rebased.

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/fx/BlackVolatilityFxProvider.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+   * <p>
+   * The returned value is a map between the parameter of a node (for example, {@code DoublesPair} specifying 
+   * expiry time and strike for a surface) and the sensitivity value to the node as {@code Double}.  
+   * 
+   * @param point  the point sensitivity at a given key
+   * @return the sensitivity to the nodes
+   */
+  public abstract Map<?, Double> nodeSensitivity(FxOptionSensitivity point);
+
+  public abstract SurfaceCurrencyParameterSensitivity surfaceParameterSensitivity(FxOptionSensitivity point);
```
Comment (marc-opengamma) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/fx/BlackVolatilityFxProvider.java]: Javadoc missing

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/fx/BlackVolatilityFxProvider.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  /**
+   * Returns the Black volatility.
+   * 
+   * @param currencyPair the currency pair
+   * @param expiryDate  the option expiry
+   * @param strike  the option strike
+   * @param forward  the underling forward
+   * @return the volatility
+   */
+  public abstract double getVolatility(CurrencyPair currencyPair, LocalDate expiryDate, double strike, double forward);
```
Comment (marc-opengamma) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/fx/BlackVolatilityFxProvider.java]: As the expiry is measured in date/time for relativeTime, it should be also for the volatility. The type of the second variable of getVolatility should probably be ZonedDateTime (in one or three variables).

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/fx/BlackVolatilityFxProvider.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+  /**
+   * Converts a date to a relative {@code double} time.
+   * 
+   * @param date  the date to find the relative time of
+   * @param time  the time to find the relative time of
+   * @param zone  the time zone
+   * @return the relative time
+   */
+  public abstract double relativeTime(LocalDate date, LocalTime time, ZoneId zone);
```
Comment (marc-opengamma) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/fx/BlackVolatilityFxProvider.java]: Change to ZonedDateTime, instead of three objects.

**Comment by jodastephen** [comment]
Branch rebased and tidied. Replaced `DoublesCurve` with `NodalCurve`.

The code refers to an analytics class - `SmileDeltaTermStructureParametersStrikeInterpolation` - that will need to be ported to Strata.

**Pull request merged**